### ISSUE TITLE
perf: lazy-import sentence-transformers and torch only when needed

### ISSUE THREAD

**Issue by ard12** [Abhigyan-Shekhar/Waggle-mcp#255]
## Problem

`sentence-transformers` (and the `torch` it pulls in) is the slowest import in the whole codebase. Cold start of `waggle-mcp` is dominated by it, even when the user has set `WAGGLE_MODEL=deterministic` and never touches the real model.

We already lazy-load the model itself, but the imports happen at module level (or close to it). Every CLI invocation eats the import cost.

## What to do

Audit imports of these three packages and push them down to function-local imports wherever they are not used at module load:

- `sentence_transformers`
- `torch`
- `numpy` (only the heavy paths; module-level is fine)

Specifically:

1. Search `src/waggle/embeddings.py` for top-level `from sentence_transformers ...` and move them inside `_load_transformer_model` (already lazy in places, verify everywhere).
2. Check `src/waggle/server.py`, `src/waggle/graph.py`, and `src/waggle/retrieval/hybrid.py` for the same pattern.
3. After the change, `waggle-mcp --help` should not import torch. Verify with:
   ```bash
   python -X importtime -c "import waggle.server" 2>&1 | grep -E "torch|sentence_transformers"
   ```

## Acceptance criteria

- `waggle-mcp --help` does not trigger `torch` or `sentence_transformers` import
- `WAGGLE_MODEL=deterministic waggle-mcp serve` does not trigger them either
- `waggle-mcp serve` (with default model) DOES trigger them on first embedding call (verify timing)
- All tests still pass
- A benchmark in the PR description showing before/after `waggle-mcp --help` time (use `python -X importtime` or `hyperfine`)

## Notes

- Be careful with type annotations: `from sentence_transformers import SentenceTransformer` for `def __init__(self) -> SentenceTransformer:` will keep the import alive. Use `TYPE_CHECKING` guards if you need the type at the boundary.
- Don't go overboard. Only lazy-import the genuinely expensive packages. Lazy-importing `os` or `json` makes the code harder to read for no win.
- This issue is `level:intermediate` because it requires a CI verification step (the importtime measurement) and a benchmark, not just a code change.

**Comment by vyshali-arc** [comment]
Hi @ard12,

I would like to work on this issue.

My plan is to:

* Audit imports of `sentence_transformers`, `torch`, and related heavy dependencies in the identified modules.
* Move expensive imports into the code paths where they are actually required while preserving existing behavior.
* Use `TYPE_CHECKING` guards where necessary to avoid eager imports caused by type annotations.
* Verify that `waggle-mcp --help` and `WAGGLE_MODEL=deterministic waggle-mcp serve` do not import `torch` or `sentence_transformers`.
* Measure and report before/after import times as requested in the PR description.
* Run the relevant test suite and CI validations before submitting the PR.

Please assign this issue to me if it is available. Thank you!

**Comment by Bhagyashri77777** [comment]
Hi @Abhigyan-Shekhar,

I would like to work on this issue under GSSoC 2026.

My approach would be to audit the imports of `sentence_transformers`, `torch`, and other heavy dependencies in the mentioned modules, move them to function-level imports where appropriate, and use `TYPE_CHECKING` guards when needed to avoid eager imports through type annotations.

I will also verify the behavior using `python -X importtime`, ensure that `waggle-mcp --help` and `WAGGLE_MODEL=deterministic waggle-mcp serve` do not trigger these imports, run the relevant tests, and include before/after benchmark results in the PR description.

Thank you for considering my request.

**Comment by SOHALIYAJAY** [comment]
Hi @Abhigyan-Shekhar ,
I would like to work on this issue, please assign me .
Thank You !!

**Comment by Monica-CodingWorld** [comment]
Hi @ard12 & @Abhigyan-Shekhar , 
I'd love to take this issue on if it's still available. Please assign it to me.

Here's how I plan to approach it:

* Review `src/waggle/embeddings.py`, `src/waggle/server.py`, `src/waggle/graph.py`, and `src/waggle/retrieval/hybrid.py` to identify eager imports of heavy dependencies such as `sentence_transformers`, `torch`, and any costly `numpy` imports.
* Move those imports into the specific functions or execution paths where they're actually needed, especially around model loading and embedding generation.
* Check type annotations to make sure they aren't unintentionally triggering imports, using `TYPE_CHECKING` guards or string annotations where appropriate.
* Verify with `python -X importtime` that commands like `waggle-mcp --help` and `WAGGLE_MODEL=deterministic waggle-mcp serve` no longer pull in `torch` or `sentence_transformers` during startup.
* Ensure the standard embedding workflow still loads the required dependencies correctly when they're first needed.
* Run the test suite to confirm there are no regressions and gather before/after startup metrics to include in the PR.
* Keep the changes focused on genuinely expensive dependencies and avoid introducing unnecessary lazy imports elsewhere in the codebase.

**Comment by ard12** [comment]
Thank you everyone for your interest in this issue.

Since @vyshali-arc was the first contributor to express interest and provide a proposed approach, I'm assigning this issue to them.

@vyshali-arc, please go ahead and work on the issue. Feel free to reach out if you need any clarification while implementing the changes.

@Bhagyashri77777, @SOHALIYAJAY, and @Monica-CodingWorld, thank you for taking the time to review the issue and share your plans. There are other open issues available in the repository, and I'd encourage you to pick up another one that interests you. We appreciate your willingness to contribute and look forward to your future PRs.

Happy contributing!

**Comment by vyshali-arc** [comment]
Hi @ard12,

I investigated the current state of the repository while working on #255.

From my testing:

* `SentenceTransformer` is already imported lazily inside `_load_transformer_model()`.
* `python -X importtime -c "import waggle.server"` does not show `torch` or `sentence_transformers`.
* The same holds when `WAGGLE_MODEL=deterministic`.

I also checked the existing tests and found coverage for deterministic fallback behavior, but I did not find import-safety tests verifying that startup paths remain free of `torch` / `sentence_transformers` imports.

Before proceeding, could you confirm whether there is any remaining code change expected for #255, or whether the intended contribution is to add import-safety tests and benchmark verification?

Thank you!

**Comment by tonmoy007** [comment]
Following up on this — I've been working on #273 (https://github.com/Abhigyan-Shekhar/Waggle-mcp/pull/273), which provides exactly the import-safety tests mentioned here.

**PR #273** adds 5 subprocess-based regression tests that verify:
- `import waggle` — ✅ no torch/sentence_transformers
- `from waggle.server import _build_parser` — ✅ no heavy imports
- `from waggle.graph import MemoryGraph` — ✅ clean
- `EmbeddingModel` with `WAGGLE_MODEL=deterministic` + background warmup — ✅ clean
- `from waggle.server import main` — ✅ clean (catches forward-reference leaks like #255 mentions)

Since @vyshali-arc confirmed the lazy-import implementation is already in place (torch/sentence_transformers don't show up on `import waggle.server`), the remaining piece of #255 is exactly these regression tests — and they're ready in #273, passing and reviewed by CodeRabbit (no actionable issues).

Would love to see if we can align these two. Perhaps the simplest path:
1. Merge #273 (import-safety tests) to close out the verification aspect of #255
2. If any remaining lazy-import refinements are still needed, those can go in a follow-up

@ard12 — wdyt?
